caffe2/onnx/onnxifi_transformer.cc

```diff
@@ -0,0 +1,371 @@
+#include "caffe2/onnx/onnxifi_transformer.h"
+
+#include <iostream>
+#include <unordered_set>
+
+#include <google/protobuf/text_format.h>
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/logging.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/core/tensor.h"
+#include "caffe2/onnx/onnx_exporter.h"
+#include "caffe2/opt/backend_cutting.h"
+
+namespace caffe2 {
+
+namespace {
+
+// TODO(yinghai): Remove the awkward conversion between unordered_map and map
+std::unordered_map<std::string, TensorShape> InferShapes(
+    Workspace* ws,
+    NetDef* pred_net,
+    CaffeMap<std::string, TensorShape>* shape_hints_ordered) {
+  // Populate shapes from workplace
+  const std::vector<string>& ws_blobs = ws->Blobs();
+  for (const auto& s : ws_blobs) {
+    shape_hints_ordered->emplace(s, GetTensorShapeOfBlob(ws->GetBlob(s)));
```

Comment:
What if blob contains sparse data rather than a tensor? We don't support such blobs in ONNX/ONNXIFI, but may still want to offload a part of the model that uses dense tensors.

Reply:
What's the impact of sparse data on the flow? It will break the shape inference as we don't support sparse tensor shape inference? In this case, we should just run the net once to get the shape info and put it in the hints, which is done in the onnxifi_transform.py. 

If sparse data results in variable sized shapes downstream, then we are in trouble as I don't know whether there is backend that supports variable sized inputs yet.

Reply:
I'm more concerned about sparse data breaking `GetTensorShapeOfBlob`. We need to handle situation where some blobs are not tensors, and just exclude producing/consuming ops from graph cutting.

Reply:
Hmm, it boils down to what ops can produce/consume sparse tensors. We can just exclude them from the subgraph if it is the case.

Reply:
As for broken `GetTensorShapeOfBlob`, it is a good catch. I added logic to skip adding shape with `unknow_shape==true` to hints.